packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts

```diff
@@ -149,25 +212,30 @@ export class FederatedQuadSource implements ILazyQuadSource {
     //       See discussion at https://github.com/comunica/comunica/issues/553
     const sourcesIt = this.sources.iterator();
     const proxyIt: AsyncIterator<PromiseProxyIterator<RDF.Quad>> = sourcesIt.map((source) => {
+      const sourceId = this.getSourceId(source);
       remainingSources++;
       sourcesCount++;
 
-      // If we can predict that the given source will have no bindings for the given pattern,
-      // return an empty iterator.
-      const pattern: Algebra.Pattern = this.algebraFactory.createPattern(
-        FederatedQuadSource.nullToVariable(subject, 's'),
-        FederatedQuadSource.nullToVariable(predicate, 'p'),
-        FederatedQuadSource.nullToVariable(object, 'o'),
-        FederatedQuadSource.nullToVariable(graph, 'g'),
-      );
+      return new PromiseProxyIterator(async () => {
+        // Convert falsy terms to variables, reverse operation of {@link ActorRdfResolveQuadPatternSource#variableToNull}.
+        // Deskolemize terms, so we send the original blank nodes to each source.
+        // Note that some sources may not match bnodes by label. SPARQL endpoints for example consider them variables.
+        const s = !subject   ? DataFactory.variable('vs') : FederatedQuadSource.deskolemizeTerm(subject,   sourceId);
+        const p = !predicate ? DataFactory.variable('vp') : FederatedQuadSource.deskolemizeTerm(predicate, sourceId);
```

Comment:
Predicates shouldn't be blanks I guess, but better to be safe than sorry.

Reply:
True. By keeping it generic, we can in the future have support for [generalized RDF](https://www.w3.org/TR/rdf11-concepts/#section-generalized-rdf) should that ever be needed.